ntoskrnl/mm/section.c

```diff
@@ -987,45 +987,43 @@ FreeSegmentPage(PMM_SECTION_SEGMENT Segment, PLARGE_INTEGER Offset)
     MmReleasePageMemoryConsumer(MC_USER, Page);
 }
 
+_When_(OldIrql == MM_NOIRQL, _IRQL_requires_max_(DISPATCH_LEVEL))
+_When_(OldIrql == MM_NOIRQL, _Requires_lock_not_held_(MmPfnLock))
+_When_(OldIrql != MM_NOIRQL, _Requires_lock_held_(MmPfnLock))
+_When_(OldIrql != MM_NOIRQL, _Releases_lock_(MmPfnLock))
+_When_(OldIrql != MM_NOIRQL, _IRQL_restores_(OldIrql))
+_When_(OldIrql != MM_NOIRQL, _IRQL_requires_(DISPATCH_LEVEL))
 VOID
 NTAPI
 MmDereferenceSegmentWithLock(PMM_SECTION_SEGMENT Segment, KIRQL OldIrql)
 {
-    BOOLEAN HaveLock = FALSE;
-
     /* Lock the PFN lock because we mess around with SectionObjectPointers */
     if (OldIrql == MM_NOIRQL)
     {
-        HaveLock = TRUE;
         OldIrql = MiAcquirePfnLock();
     }
 
     if (InterlockedDecrement64(Segment->ReferenceCount) > 0)
     {
         /* Nothing to do yet */
-        if (HaveLock)
-            MiReleasePfnLock(OldIrql);
+        MiReleasePfnLock(OldIrql);
         return;
     }
 
     *Segment->Flags |= MM_SEGMENT_INDELETE;
-
-    if (HaveLock)
-        MiReleasePfnLock(OldIrql);
+    MiReleasePfnLock(OldIrql);
```

Comment:
Repeating my other comment here: Just because the lock is released by all callers right after the call to this function is not argument to do it here, because it doesn't necessarily have to stay like that. It's also very unintuitive behavior. IMO no function, that isn't explicitly called "AcquireLock" or "ReleaseLock" or alike should acquire or release a lock. I'm already horrified of some Windows function names, that are called like "CheckForX", but then change X and do all kinds of stuff, no need to introduce more of that.